active_learning/learn.py

```diff
@@ -0,0 +1,136 @@
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.linear_model import SGDClassifier
+from sklearn.metrics import confusion_matrix, precision_recall_fscore_support, accuracy_score
+
+from libact.base.dataset import Dataset
+from libact.query_strategies import *
+from libact.models import *
+
+from active_learning.models import Article, Label, DataSet
+from active_learning.libact_models import *
+
+
+def train_active_learning(num_queries, training_dataset, oracle_training_dataset):
+    active_learning_strategy = 1
+    query_strategy, strategy = get_active_learning_strategy(active_learning_strategy, training_dataset)
+    model = SGDClassifier(loss='hinge', penalty='l2', alpha=1e-2, n_iter=1000, random_state=None, learning_rate='optimal', class_weight='balanced')
+    # train on labeled instances
+    model.fit(*(training_dataset.format_sklearn()))
+    # active learning
+    for query_num in range(num_queries):
+        # query the active learning strategy
+        query_id = query_strategy.make_query()
+        # use the oracle to label the query
+        label = oracle_training_dataset.data[query_id][1]
+        # update dataset with label then retrain the model
+        training_dataset.update(query_id, label)
+        model.fit(*(training_dataset.format_sklearn()))
+    return model
+
+
+def cross_validate_active_learning(num_folds, X, y, labeled_instances, num_queries):
+    results = []
+    # make K folds
+    skf = StratifiedKFold(n_splits=num_folds)
+    skf.get_n_splits(X, y)
+    # for each fold
+    for train_index, test_index in skf.split(X, y):
+        X_train, X_test = X[train_index], X[test_index]
+        y_train, y_test = y[train_index], y[test_index]
+        # hide all but labeled_instances in training set
+        X_train, X_train_unlabeled, y_train, y_train_oracle = train_test_split(X_train, y_train, train_size=labeled_instances, random_state=None, stratify=y_train)
+        # join X to obtain semi-labeled dataset
+        X_train = numpy.array(vstack([X_train, X_train_unlabeled]).toarray())
+        # fill in y_train with missing labels for active learning on semi-labeled dataset
+        y_train = y_train.tolist()
+        y_train_with_missing = y_train + [None] * len(y_train_oracle)
+        # build dataset object for libact library
+        training_dataset = Dataset(X_train, y_train_with_missing)
+        oracle_training_dataset = Dataset(X_train, y_train + y_train_oracle.tolist())
+        # do active_learning and get trained model
+        trained_model = train_active_learning(num_queries, training_dataset, oracle_training_dataset)
+        # test model
+        predicted = trained_model.predict(X_test)
+        results.append({'y_true': y_test, 'y_pred': predicted})
+    return results
+
+
+def preprocess(X_strings):
+    return numpy.array(TfidfVectorizer(stop_words='english').fit_transform(X_strings).toarray())
+
+
+def learn():
+    # Build dataset
+    dataset_name = 'Four University'
```

Comment:
Should be a parameter to be able to run on other datasets. Same with other variables below.